File: map/app-form/src/app/services/ods-table/ods.ts
import * as XLSX from 'xlsx'

function columnIdx (rowLetter) {
    // {:pre [(char? rowLetter)]} // TODO define contract
    return rowLetter
}

function getRowCount(sheet) {
    return XLSX.utils.sheet_to_json(sheet).length
}

function identity(arg: any): any {
    return arg;
}

function range(n) {
    return Array.from(Array(n).keys())
}

function sheetRowIndexes(sheet) {
  const rowCount = getRowCount(sheet)
  const arrSheetRange = range(rowCount)
  // sheet line numbers start from 1 and we need to skip the header
  const shiftedArr = arrSheetRange.map((v:number) => v + 2)
  return shiftedArr
}

function textAtPosition(sheet, p, row) {
    const cellAddr = p+row
    const cell = sheet[cellAddr];
    return (cell ? cell.v : undefined);
}

function address(sheet, row) {
    return textAtPosition(sheet, columnIdx('C'), row)
}

function address_receiver(sheet, row) {
    return textAtPosition(sheet, columnIdx('B'), row)
}

function rowNr(idx) {
    return idx + 2
}

export function replaceAll(s: string, match: string, replacement: string): string {
    // https://www.designcise.com/web/tutorial/how-to-replace-all-occurrences-of-a-word-in-a-javascript-string
    if (s) {
        return s.replace(new RegExp(match, 'g'), replacement)
    }
    else {
        return s
    }
}

// function composition
// https://dev.to/ascorbic/creating-a-typed-compose-function-in-typescript-3-351i
export const pipe = <T extends any[], R>(
    fn1: (...args: T) => R,
    ...fns: Array<(a: R) => R>
) => {
    const piped = fns.reduce(
        (prevFn, nextFn) => (value: R) => nextFn(prevFn(value)),
        value => value
    );
    return (...args: T) => piped(fn1(...args));
};

export const compose = <R>(fn1: (a: R) => R, ...fns: Array<(a: R) => R>) =>
    fns.reduce((prevFn, nextFn) => value => prevFn(nextFn(value)), fn1);

function cleanup(a) {
    const f1 = (a: string) => replaceAll(a, ' ', ' ')
    const f2 = (a: string) => replaceAll(a, '  ', ' ')
    const f3 = (a: string) => replaceAll(a, '  ', ' ')
    // repeating twice should be enough; simple and dirty
    const f4 = (a: string) => replaceAll(a, ' \n', '\n')
    const f5 = (a: string) => replaceAll(a, ' \n', '\n')
    const f6 = (a: string) => a.trim()
    // e.V. without space is incorrect
    // const f7 = (a: string) =>replaceAll(a, 'e. V.', 'e.V.')
    const pipedFunction = pipe(f1, f2, f3, f4, f5, f6);
    if (typeof a !== 'undefined') {
        return pipedFunction(a)
    }
    // else {
    //     return s
    // }
    return null
}

export interface IHash {
    [details: string] : string;
}


// hashmap keywords start with '_'
export const _idx = '_idx'
export const _address ='_address'

function calcAddresses(sheet, rows) {
    const f1 = (a: string) => replaceAll(a, '\n', ', ')
    const f2 = cleanup
    const f3 = (row: number) => address(sheet, row)
    const pipedFunction = pipe(f3, f2, f1)

    const vals = rows.map(pipedFunction)
    return vals.map((s, i) => {
        let hm: IHash = {};
        hm[_idx] = rowNr(i)
        hm[_address] = s
        return hm
    })
}

export const _addr_recv ='_addr_recv'
function calcAddressReceiver(sheet, rows) {
    const f1 = (a: string) => replaceAll(a, '\n', ', ')
    const f2 = cleanup
    const f3 = (row: number) => address_receiver(sheet, row)
    const pipedFunction = pipe(f3, f2, f1)

    const vals = rows.map(pipedFunction)
    return vals.map((s, i) => {
        let hm: IHash = {};
        hm[_idx] = rowNr(i)
        hm[_addr_recv] = s
        return hm
    })
}

function association(sheet, row) {
    return textAtPosition(sheet, columnIdx('A'), row)
}

export const _name = '_name'
function calcAssociations(sheet, rows) {
    const f1 = (a: string) => replaceAll(a, '\n', ' ')
    const f2 = cleanup
    const f3 = (row: number) => association(sheet, row)
    const pipedFunction = pipe(f3, f2, f1)

    const vals = rows.map(pipedFunction)
    return vals.map((s, i) => {
        let hm: IHash = {};
        hm[_idx] = rowNr(i)
        hm[_name] = s
        return hm
    })
}

function cityDistrict(sheet, row) {
    return textAtPosition(sheet, columnIdx('D'), row)
}

export const _cityDistrict = '_cityDistrict'
function calcDistricts(sheet, rows) {
    const f1 = identity
    const f2 = cleanup
    const f3 = (row: number) => cityDistrict(sheet, row)
    const pipedFunction = pipe(f3, f2, f1)

    const vals = rows.map(pipedFunction)
    return vals.map((s, i) => {
        let hm: IHash = {};
        hm[_idx] = rowNr(i)
        hm[_cityDistrict] = s
        return hm
    })
}

function tableCoordinates(sheet, row) {
    return textAtPosition(sheet, columnIdx('E'), row)
}

export const _coordinates = '_coordinates'
function calcCoordinates(sheet, rows) {
    const f1 = identity
    const f2 = cleanup
    const f3 = (row: number) => tableCoordinates(sheet, row)
    const pipedFunction = pipe(f3, f2, f1)

    const vals = rows.map(pipedFunction)
    return vals.map((s, i) => {
        let hm: IHash = {};
        hm[_idx] = rowNr(i)
        hm[_coordinates] = s
        return hm
    })
}

function contact(sheet, row) {
    return textAtPosition(sheet, columnIdx('F'), row)
}

export const _contact = '_contact'
function calcContacts(sheet, rows) {
    const f1 = identity
    const f2 = cleanup
    const f3 = (row: number) => contact(sheet, row)
    const pipedFunction = pipe(f3, f2, f1)

    const vals = rows.map(pipedFunction)
    return vals.map((s, i) => {
        let hm: IHash = {};
        hm[_idx] = rowNr(i)
        hm[_contact] = s
        return hm
    })
}

function logo(sheet, row) {
    return textAtPosition(sheet, columnIdx('G'), row)
}

export const _logo = '_logo'
function calcLogos(sheet, rows) {
    const f1 = identity
    const f2 = cleanup
    const f3 = (row: number) => logo(sheet, row)
    const pipedFunction = pipe(f3, f2, f1)

    const vals = rows.map(pipedFunction)
    return vals.map((s, i) => {
        let hm: IHash = {};
        hm[_idx] = rowNr(i)
        hm[_logo] = s
        return hm
    })
}

function webPage(sheet, row) {
    return textAtPosition(sheet, columnIdx('H'), row)
}

export const _webPage = '_webPage'
function calcWebPages(sheet, rows) {
    const f1 = identity
    const f2 = cleanup
    const f3 = (row: number) => webPage(sheet, row)
    const pipedFunction = pipe(f3, f2, f1)

    const vals = rows.map(pipedFunction)
    return vals.map((s, i) => {
        let hm: IHash = {};
        hm[_idx] = rowNr(i)
        hm[_webPage] = s
        return hm
    })
}

function goal(sheet, row) {
    return textAtPosition(sheet, columnIdx('I'), row)
}

export const _goal = '_goal'
function calcGoals(sheet, rows) {
    const f1 = identity
    const f2 = cleanup
    const f3 = (row: number) => goal(sheet, row)
    const pipedFunction = pipe(f3, f2, f1)

    const vals = rows.map(pipedFunction)
    return vals.map((s, i) => {
        let hm: IHash = {};
        hm[_idx] = rowNr(i)
        hm[_goal] = s
        return hm
    })
}

function activity(sheet, row) {
    return textAtPosition(sheet, columnIdx('J'), row)
}

export const _activity = '_activity'
function calcActivities(sheet, rows) {
    const f1 = identity
    const f2 = cleanup
    const f3 = (row: number) => activity(sheet, row)
    const pipedFunction = pipe(f3, f2, f1)

    const vals = rows.map(pipedFunction)
    return vals.map((s, i) => {
        let hm: IHash = {};
        hm[_idx] = rowNr(i)
        hm[_activity] = s
        return hm
    })
}

export const _desc = 'desc'
export function calcReadTable(sheet) {
  const rowIndexes   = sheetRowIndexes(sheet)
  const associations = calcAssociations(sheet, rowIndexes)
  const addresses    = calcAddresses(sheet, rowIndexes)
  const addr_recv    = calcAddressReceiver(sheet, rowIndexes)
  const districts    = calcDistricts(sheet, rowIndexes)
  const contacts     = calcContacts(sheet, rowIndexes)
  const webPages     = calcWebPages(sheet, rowIndexes)
  const goals        = calcGoals(sheet, rowIndexes)
  const activities   = calcActivities(sheet, rowIndexes)
  const coordinates  = calcCoordinates(sheet, rowIndexes)
  const logos        = calcLogos(sheet, rowIndexes)

  let mm = new Array()

  // poor man's merge
  for (let y in associations) {
    // console.log(
    //         'y: ' + y +
    //         '; _idx: ' + associations[y][_idx] +
    //         '; _name: ' + associations[y][_name] +
    //         '; _address: '+addresses[y][_address]
    // )

    let m: IHash = {};
    m[y]             = y
    m[_idx]          = associations[y][_idx]
    m[_name]         = associations[y][_name]
    m[_address]      = addresses[y][_address]
    m[_addr_recv]    = addr_recv[y][_addr_recv]
    m[_cityDistrict] = districts[y][_cityDistrict]
    m[_desc]         = (contacts[y][_contact] + "\n\n" + webPages[y][_webPage]).trim()
    m[_coordinates]  = coordinates[y][_coordinates]
    m[_contact]      = contacts[y][_contact]
    m[_logo]         = logos[y][_logo]
    m[_webPage]      = webPages[y][_webPage]
    m[_goal]         = goals[y][_goal]
    m[_activity]     = activities[y][_activity]
    mm.push(m)
  }
  return mm
}

File: src/clj/fdk/data/ods.clj
(ns fdk.data.ods
  (:require
   [clojure.tools.logging :refer :all]
   [fdk.data.utils :refer :all]
   [clojure.string :as cstr]
   [djy.char :as djy]
   ;; [clojure.inspector :refer :all]
   ;; [taoensso.timbre :as timbre :refer []]
   )
  (:import org.odftoolkit.simple.SpreadsheetDocument))

(defonce cache (atom {}))

(defn cache!
  "Also return the cached value for further consumption."
  [calc-data-fn ks]
  (let [data (calc-data-fn)]
    #_(swap! cache update-in ks (fn [_] data))
    data))

(defn- column-idx [row-letter]
  {:pre [(char? row-letter)]}
  (.indexOf (djy/char-range \A \Z) (djy/upper-case row-letter)))

(defn sheet0 [fname] (-> fname
                         (SpreadsheetDocument/loadDocument)
                         (.getSheetByIndex 0)))

(defn sheet-rows [sheet]
  (drop 1 ;; skip column header
        (
         identity
         ;; drop 6
         (
          identity
          ;; take 8
          (range (.getRowCount sheet))))))

(defn text-at-position [sheet p row]
  (.getDisplayText (.getCellByPosition sheet p row)))

(defn address [sheet row] (text-at-position sheet (column-idx \C) row))

(defn row-nr [idx] (+ idx 2))

(defn- cleanup [a]
  ((apply comp (reverse [
                         cstr/trim
                         ;; repeating twice should be enough; simple and dirty
                         (fn [a] (cstr/replace a " \n" "\n"))
                         (fn [a] (cstr/replace a " \n" "\n"))
                         (fn [a] (cstr/replace a "  " " "))
                         (fn [a] (cstr/replace a "  " " "))
                         (fn [a] (cstr/replace a " " " "))
                         ;; e.V. without space is incorrect
                         ;; (fn [a] (cstr/replace a "e. V." "e.V."))
                         ]))
   a))

(defn calc-addresses
  "A list of indexed hash-maps:
  '({:idx 0 :address \"...\"}
    {:idx 1 :address \"...\"}
    {:idx 2 :address \"...\"})"
  [sheet]
  (->> sheet
       (sheet-rows)
       (map (comp
             (fn [a] (cstr/replace a "\n" ", "))
             cleanup
             (partial address sheet)))
       (map-indexed (fn [i s] {:idx (row-nr i) :address s}))))

(defn addresses [sheet]
  (let [ks [:addresses]]
    (if-let [v (get-in @cache ks)]
      v
      (cache! (fn [] (calc-addresses sheet)) ks))))

(defn association [sheet row] (text-at-position sheet (column-idx \A) row))

(defn calc-associations
  "A list of indexed hash-maps:
  '({:idx 0 :name \"...\"}
    {:idx 1 :name \"...\"}
    {:idx 2 :name \"...\"})"
  [sheet]
  (->> sheet
       (sheet-rows)
       (map (comp
             (fn [a] (cstr/replace a "\n" " "))
             cleanup
             (partial association sheet)))
       (map-indexed (fn [i s] {:idx (row-nr i) :name s}))))

(defn associations [sheet]
  (let [ks [:associations]]
    (if-let [v (get-in @cache ks)]
      v
      (cache! (fn [] (calc-associations sheet)) ks))))

(defn city-district [sheet row] (text-at-position sheet (column-idx \D) row))

(defn calc-city-districts
  "A list of indexed hash-maps:
  '({:idx 0 :city-district \"...\"}
    {:idx 1 :city-district \"...\"}
    {:idx 2 :city-district \"...\"})"
  [sheet]
  (->> sheet
       (sheet-rows)
       (map (comp
             cleanup
             (partial city-district sheet)))
       (map-indexed (fn [i s] {:idx (row-nr i) :city-district s}))))

(defn city-districts [sheet]
  (let [ks [:city-districts]]
    (if-let [v (get-in @cache ks)]
      v
      (cache! (fn [] (calc-city-districts sheet)) ks))))

(defn table-coordinates [sheet row] (text-at-position sheet (column-idx \E) row))

(defn calc-coordinates
  "A list of indexed hash-maps:
  '({:idx 0 :coordinates \"...\"}
    {:idx 1 :coordinates \"...\"}
    {:idx 2 :coordinates \"...\"})"
  [sheet]
  (->> sheet
       (sheet-rows)
       (map (comp
             cleanup
             (partial table-coordinates sheet)))
       (map-indexed (fn [i s] {:idx (row-nr i) :coordinates s}))))

(defn coordinates [sheet]
  (let [ks [:coordinates]]
    (if-let [v (get-in @cache ks)]
      v
      (cache! (fn [] (calc-coordinates sheet)) ks))))

(defn contact [sheet row] (text-at-position sheet (column-idx \F) row))

(defn calc-contacts
  "A list of indexed hash-maps:
  '({:idx 0 :contact \"...\"}
    {:idx 1 :contact \"...\"}
    {:idx 2 :contact \"...\"})"
  [sheet]
  ((comp
    #_(fn [s] (println "1" s) s)
    (partial map-indexed (fn [i s] {:idx (row-nr i) :contact s}))
    (partial map (comp
                  cleanup
                  #_(fn [s] (println "0" s) s)
                  (partial contact sheet)))
    sheet-rows)
   sheet))

(defn contacts [sheet]
  (let [ks [:contacts]]
    (if-let [v (get-in @cache ks)]
      v
      (cache! (fn [] (calc-contacts sheet)) ks))))

(defn logo [sheet row]
  #_(println "logos" (text-at-position sheet (column-idx \G) row))
  (text-at-position sheet (column-idx \G) row))

(defn calc-logos
  "A list of indexed hash-maps:
  '({:idx 0 :web-page \"...\"}
    {:idx 1 :web-page \"...\"}
    {:idx 2 :web-page \"...\"})"
  [sheet]
  (->> sheet
       (sheet-rows)
       (map (comp
             (fn [urls]
               #_(println "(cstr/split-lines urls)" (cstr/split-lines urls))
               (cstr/split-lines urls))
             cleanup
             (partial logo sheet)))
       (map-indexed (fn [i a] {:idx (row-nr i) :logos a}))))

(defn logos [sheet]
  (let [ks [:web-pages]]
    (if-let [v (get-in @cache ks)]
      v
      (cache! (fn [] (calc-logos sheet)) ks))))

(defn web-page [sheet row] (text-at-position sheet (column-idx \H) row))

(defn calc-web-pages
  "A list of indexed hash-maps:
  '({:idx 0 :web-page \"...\"}
    {:idx 1 :web-page \"...\"}
    {:idx 2 :web-page \"...\"})"
  [sheet]
  (->> sheet
       (sheet-rows)
       (map (comp
             cleanup
             (partial web-page sheet)))
       (map-indexed (fn [i a] {:idx (row-nr i) :web-page a}))))

(defn web-pages [sheet]
  (let [ks [:web-pages]]
    (if-let [v (get-in @cache ks)]
      v
      (cache! (fn [] (calc-web-pages sheet)) ks))))

(defn goal
  "umbenannt auf Ziele des Vereins"
  [sheet row] (text-at-position sheet (column-idx \I) row))

(defn calc-goals
  "
  TODO Natural Language Understanding (NLU)
  See https://github.huggingface/transformers
  Or just count word frequency

  A list of indexed hash-maps:
  '({:idx 0 :goal \"...\"}
    {:idx 1 :goal \"...\"}
    {:idx 2 :goal \"...\"})"
  [sheet]
  (->> sheet
       (sheet-rows)
       (map (comp
             cleanup
             (partial goal sheet)))
       (map-indexed (fn [i s] {:idx (row-nr i) :goal s}))))

(defn goals [sheet]
  (let [ks [:goals]]
    (if-let [v (get-in @cache ks)]
      v
      (cache! (fn [] (calc-goals sheet)) ks))))

(defn activity
  [sheet row] (text-at-position sheet (column-idx \J) row))

(defn calc-activities
  "
  TODO Natural Language Understanding (NLU)
  See https://github.huggingface/transformers
  Or just count word frequency

  A list of indexed hash-maps:
  '({:idx 0 :activity \"...\"}
    {:idx 1 :activity \"...\"}
    {:idx 2 :activity \"...\"})"
  [sheet]
  (->> sheet
       (sheet-rows)
       (map (comp
             cleanup
             (partial activity sheet)))
       (map-indexed (fn [i s] {:idx (row-nr i) :activity s}))))

(defn activities [sheet]
  (let [ks [:goals]]
    (if-let [v (get-in @cache ks)]
      v
      (cache! (fn [] (calc-activities sheet)) ks))))

(defn calc-read-table
  "A list of indexed hash-maps:
  [{:idx 0 :name \"...\" :address \"...\" :desc \"...\"}
   {:idx 1 :name \"...\" :address \"...\" :desc \"...\"}
   {:idx 2 :name \"...\" :address \"...\" :desc \"...\"}]"
  [fname]
  ;; TODO make sure the associations and addresses are:
  ;; 1. sorted and 2. of the same size
  (let [s0 (sheet0 fname)]
    (mapv (fn [associations addresses cdistricts
              contacts web-pages goals activities
              coordinates logos]
            (let [contact (:contact contacts)
                  web-page (:web-page web-pages)]
              (merge associations
                     addresses
                     goals
                     cdistricts
                     {:desc (cstr/trim (cstr/join "\n\n" [contact web-page]))}
                     contacts
                     activities
                     coordinates
                     logos)))
          (associations s0)
          (addresses s0)
          (city-districts s0)
          (contacts s0)
          (web-pages s0)
          (goals s0)
          (activities s0)
          (coordinates s0)
          (logos s0))))

(defn read-table [fname]
  (let [fname (new java.io.File
                   "./resources/karte.ods"
                   ;; The ö is not properly encoded
                   ;; "./resources/Vereinsinformationen_öffentlich_Stadtteilkarte.ods"
                   )
        ks [:table]]
    (if-let [v (get-in @cache ks)]
      v
      (cache! (fn [] (calc-read-table fname)) ks))))

(def default-category "Sonstiges")

(defn reset-cache!
  "(fdk.data.ods/reset-cache! \"resources/Vereinsinformationen_öffentlich_Stadtteilkarte.ods\")"
  [fname]
  (swap! cache (fn [_] {}))
  (let [tbeg (System/currentTimeMillis)]
    ;; enforce evaluation; can't be done by (force (all-rankings))
    (dorun
     (calc-read-table fname))
    (printf "%s chars cached in %s ms"
            (count (str @cache)) (- (System/currentTimeMillis) tbeg))))

(defn process-table-row [districts
                              ;; : DropdownOption[]
                              row
                              ]

  ;; (println "[process-table-row]" "row" row)
  ;; (println "[process-table-row]" "(keys row)" (keys row))
  (let [association-id (str (java.util.UUID/randomUUID))]
    ;; (println "[process-table-row]" "association-id" association-id)
    ((comp
      ;; (fn [m] (println "[process-table-row]" "m" m) m)
      ;; (fn [m] (println "" m) m)
      ;; (fn [m] (println "9" m) m)
      (fn [m] (assoc m :id association-id))
      ;; (fn [m] (println "8" m) m)
      (fn [m] (assoc m :name (-> row :name)))
      ;; shortName addr_recv ;; not uses
      ;; (fn [m] (println "7" m) m)
      (fn [m] (assoc m :goals
                     {:format "plain" ;; "plain" | "html"
                      :text (-> row :goal escape-html-with-null)}))
      ;; (fn [m] (println "6" m) m)
      (fn [m] (assoc m :activities
                     {:format "plain" ;; "plain" | "html"
                      :text (-> row :activity escape-html-with-null)}))
      ;; (fn [m] (println "5" m) m)
      (fn [m] (assoc m :contacts
                     ((comp
                       (partial process-contact association-id)
                       :desc)
                      row)))
      ;; (fn [m] (println "4" m) m)
      (partial conj
               ((comp
                 ;; (fn [m] (println "31" m) m)
                 (partial process-socialMedia-links association-id)
                 ;; (fn [m] (println "30" m) m)
                 :links)
                row))
      ;; (fn [m] (println "3" m) m)
      (fn [m] (assoc m :images
                     ((comp
                       ;; (fn [m] (println "21" m) m)
                       (partial process-images association-id)
                       ;; (fn [m] (println "20" m) m)
                       :logos)
                      row)))
      ;; (fn [m] (println "2" m) m)
      (fn [m] (assoc m :districts
                     (keywords (:cityDistrict row) districts)))
      ;; (fn [m] (println "1" m) m)
      (partial conj (convert-address (-> row :address normalize-address))) ;; outputAddress
      ;; (fn [m] (println "0" m) m)
      (partial conj (-> row :coordinates parse-lat-lon))
      )
     {})))
